### PR TITLE
[docs/reference] more fixes in Markdown files

### PR DESCRIPTION
**Pull request by michelou**
Follow up of PR #10767, PR #10826 and PR #10860.

> **Note:** To be discussed with other members of the Scala core team/[Scala Center](https://scala.epfl.ch/): in his [comment](https://github.com/lampepfl/dotty/pull/10860#issuecomment-748536518) to my PR #10860 @SethTisue expresses the wish to have a more "official" location for the PDF document [scala3_reference.pdf](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf) I generate to check my changes to the Scala 3 reference documenation. Without any reaction from you (@liufengyun ), I will suggest @SethTisue to use the current link (which is stable since the creation of my GitHub account).

Fixes more issues found in Markdown files in [`docs/reference/`](https://github.com/lampepfl/dotty/tree/master/docs/docs/reference), e.g,
- removed/added spaces/lines
- corrected misspellings (e.g "covarant" -> "covariant")
- corrected left margin in code block
- reduced line width in several code blocks (to avoid text overflow in the right margin)
- added several external links (e.g. Coq)

**PS.** The above changes can also be viewed in the PDF document [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf) (841 Kb) generated with [Pandoc 2.11](https://pandoc.org/releases.html).

### CODE REVIEW

File: docs/docs/reference/dropped-features/delayed-init.md [hunk 1]
```diff
@@ -19,8 +18,9 @@ some JVM's means that it will only be interpreted. So, better not use it
...
-object Hello {
-  def main(args: Array[String]) =
-    println(s"Hello, ${args(0)}")
-}
+object Hello:
+   def main(args: Array[String]) =
+      println(s"Hello, ${args(0)}")
 ```
+On the other hand, Scala 3 offers a convenient alternative to such "program" objects
+with [@main methods](../changed-features/main-functions.html).
```
Comment (liufengyun) [inline docs/docs/reference/dropped-features/delayed-init.md:26]: Nice addition 👍 

Link to markdown file for consistency?

File: docs/docs/reference/metaprogramming/erased-terms-spec.md [hunk 2]
```diff
@@ -3,9 +3,9 @@ layout: doc-page
 title: "Erased Terms Spec"
 ---
 
-# Implementation
+## Implementation
 
-## Rules
+### Rules
```
Comment (liufengyun) [inline docs/docs/reference/metaprogramming/erased-terms-spec.md:8]: Maybe remove `Implementation` and only keep `Rules`?

File: docs/docs/reference/metaprogramming/simple-smp.md [hunk 3]
```diff
@@ -123,9 +123,9 @@ Proof by structural induction over terms.
 
 To prove (1):
 
- - the cases for variables, lambdas and applications are as in STL.
+ - the cases for variables, lambdas and applications are as in [STL](https://en.wikipedia.org/wiki/Simply_typed_lambda_calculus).
```
Comment (liufengyun) [inline docs/docs/reference/metaprogramming/simple-smp.md:126]:
```suggestion
 - the cases for variables, lambdas and applications are as in [STLC](https://en.wikipedia.org/wiki/Simply_typed_lambda_calculus).
```

**Review by liufengyun — APPROVED** [review]
LGTM, thanks @michelou 👍 

Rebase and conflict fixes are needed.

I somehow missed this comment:

> Note: To be discussed with other members of the Scala core team/Scala Center: in his comment to my PR #10860 @SethTisue expresses the wish to have a more "official" location for the PDF document scala3_reference.pdf I generate to check my changes to the Scala 3 reference documenation. Without any reaction from you (@liufengyun ), I will suggest @SethTisue to use the current link (which is stable since the creation of my GitHub account).

The difficulty is how to keep the PDF up to date with the docs. The best approach seems to be update the CI to generate the PDF as part of the website update.

File: docs/docs/reference/dropped-features/delayed-init.md (hunk 1 above)
Comment (michelou) [reply docs/docs/reference/dropped-features/delayed-init.md]: It's @odersky  addition (visible before rebasing with upstream).

File: docs/docs/reference/metaprogramming/erased-terms-spec.md (hunk 2 above)
Comment (michelou) [reply docs/docs/reference/metaprogramming/erased-terms-spec.md]: Done.

File: docs/docs/reference/metaprogramming/simple-smp.md (hunk 3 above)
Comment (michelou) [reply docs/docs/reference/metaprogramming/simple-smp.md]: Done.

File: docs/docs/reference/contextual/derivation-macro.md [hunk 4]
```diff
@@ -84,20 +78,19 @@ Instead we extract the tuple-type for element types using pattern matching over
 quotes and more specifically of the refined type:
 
 ```scala
- case '{ $m: Mirror.ProductOf[T] { type MirroredElemTypes = elementTypes }} => ...
+   case '{ $m: Mirror.ProductOf[T] { type MirroredElemTypes = elementTypes }} => ...
```
[the remark this answers is not in the document]
Comment (michelou) [reply docs/docs/reference/contextual/derivation-macro.md]: Not mine (visible before rebasing with upstream).

**Comment by michelou** [comment]
> LGTM, thanks @michelou 👍
> 
> Rebase and conflict fixes are needed.
> 
> I somehow missed this comment:
> 
> > Note: To be discussed with other members of the Scala core team/Scala Center: in his comment to my PR #10860 @SethTisue expresses the wish to have a more "official" location for the PDF document scala3_reference.pdf I generate to check my changes to the Scala 3 reference documenation. Without any reaction from you (@liufengyun ), I will suggest @SethTisue to use the current link (which is stable since the creation of my GitHub account).
> 
> The difficulty is how to keep the PDF up to date with the docs. The best approach seems to be update the CI to generate the PDF as part of the website update.

@liufengyun For me the relevant point is to know what is your position about generating/hosting a PDF version of the Scala 3 documentation, i.e. the reference documentation in this case. The technical aspects mostly reduce to my choice to use Pandoc for generating the PDF document. As a reminder I'm running [Pandoc](https://pandoc.org/releases.html) on Windows 10 with 3 additional files:
- `scala3_reference.yaml` (Pandoc config)
- `scala3_reference.md` (document structure, including title page)
- `scala3_reference.latex` (latex template, derived from Pandoc default latex template)

**Note**: I could easily write the corresponding shell script (assuming Pandoc 2.11 is available on the CI machine).

**Comment by liufengyun** [comment]
@michelou I think it's valuable to have a PDF version and make it available on the dotty website.

**Comment by michelou** [comment]
> @michelou I think it's valuable to have a PDF version and make it available on the dotty website.

@liufengyun I suggest we discuss the details per email.

**Pull request merged**